File: api/handlers/schedule.go
package handlers

import (
	"context"

	"github.com/gofiber/fiber/v2"
	"github.com/uzzalhcse/crawlify/internal/healthcheck"
	"github.com/uzzalhcse/crawlify/internal/healthcheck/notifications"
	"github.com/uzzalhcse/crawlify/internal/logger"
	"github.com/uzzalhcse/crawlify/internal/storage"
	"github.com/uzzalhcse/crawlify/pkg/models"
	"go.uber.org/zap"
)

// ScheduleHandler handles health check schedule endpoints
type ScheduleHandler struct {
	scheduleRepo     *storage.HealthCheckScheduleRepository
	schedulerService *healthcheck.SchedulerService
}

// NewScheduleHandler creates a new schedule handler
func NewScheduleHandler(
	scheduleRepo *storage.HealthCheckScheduleRepository,
	schedulerService *healthcheck.SchedulerService,
) *ScheduleHandler {
	return &ScheduleHandler{
		scheduleRepo:     scheduleRepo,
		schedulerService: schedulerService,
	}
}

// GetSchedule retrieves the schedule for a workflow
func (h *ScheduleHandler) GetSchedule(c *fiber.Ctx) error {
	workflowID := c.Params("id")
	ctx := context.Background()

	schedule, err := h.scheduleRepo.GetByWorkflowID(ctx, workflowID)
	if err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"error": "Schedule not found",
		})
	}

	return c.JSON(schedule)
}

// CreateSchedule creates or updates a health check schedule
func (h *ScheduleHandler) CreateSchedule(c *fiber.Ctx) error {
	workflowID := c.Params("id")
	ctx := context.Background()

	var req struct {
		Schedule           string                     `json:"schedule"`
		Enabled            bool                       `json:"enabled"`
		NotificationConfig *models.NotificationConfig `json:"notification_config,omitempty"`
	}

	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Invalid request body",
		})
	}

	// Check if schedule already exists
	existingSchedule, err := h.scheduleRepo.GetByWorkflowID(ctx, workflowID)
	if err == nil && existingSchedule != nil {
		// Update existing
		existingSchedule.Schedule = req.Schedule
		existingSchedule.Enabled = req.Enabled
		existingSchedule.NotificationConfig = req.NotificationConfig

		err = h.scheduleRepo.Update(ctx, existingSchedule)
		if err != nil {
			logger.Error("Failed to update schedule", zap.Error(err))
			return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
				"error": "Failed to update schedule",
			})
		}

		return c.JSON(existingSchedule)
	}

	// Create new
	schedule, err := h.schedulerService.CreateSchedule(ctx, workflowID, req.Schedule, req.NotificationConfig)
	if err != nil {
		logger.Error("Failed to create schedule", zap.Error(err))
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to create schedule",
		})
	}

	return c.Status(fiber.StatusCreated).JSON(schedule)
}

// DeleteSchedule deletes a health check schedule
func (h *ScheduleHandler) DeleteSchedule(c *fiber.Ctx) error {
	workflowID := c.Params("id")
	ctx := context.Background()

	err := h.scheduleRepo.Delete(ctx, workflowID)
	if err != nil {
		logger.Error("Failed to delete schedule", zap.Error(err))
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Failed to delete schedule",
		})
	}

	return c.SendStatus(fiber.StatusNoContent)
}

// TestNotification sends a test notification
func (h *ScheduleHandler) TestNotification(c *fiber.Ctx) error {
	workflowID := c.Params("id")

	var req struct {
		NotificationConfig *models.NotificationConfig `json:"notification_config"`
	}

	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Invalid request body",
		})
	}

	// Create a dummy report for testing
	testReport := &models.HealthCheckReport{
		ID:           "test-notification",
		WorkflowID:   workflowID,
		WorkflowName: "Test Workflow",
		Status:       models.HealthCheckStatusHealthy,
		Summary: &models.HealthCheckSummary{
			TotalNodes:   5,
			PassedNodes:  5,
			FailedNodes:  0,
			WarningNodes: 0,
		},
	}

	// Send test notification
	if req.NotificationConfig != nil && req.NotificationConfig.Slack != nil {
		slackNotifier := notifications.NewSlackNotifier()
		err := slackNotifier.Send(req.NotificationConfig.Slack, testReport, "Test Workflow")
		if err != nil {
			logger.Error("Failed to send test notification", zap.Error(err))
			return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
				"error": "Failed to send notification: " + err.Error(),
			})
		}
	}

	return c.JSON(fiber.Map{
		"message": "Test notification sent successfully",
	})
}

File: frontend/src/stores/schedule.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { workflowsApi } from '@/api/workflows'
import type { HealthCheckSchedule, NotificationConfig } from '@/types'

export const useScheduleStore = defineStore('schedule', () => {
    const schedules = ref<Map<string, HealthCheckSchedule>>(new Map())
    const loading = ref(false)
    const error = ref<string | null>(null)

    // Get schedule for a workflow
    const getSchedule = computed(() => {
        return (workflowId: string) => schedules.value.get(workflowId)
    })

    // Fetch schedule from API
    async function fetchSchedule(workflowId: string) {
        loading.value = true
        error.value = null
        try {
            const response = await workflowsApi.getSchedule(workflowId)
            schedules.value.set(workflowId, response.data)
            return response.data
        } catch (err: any) {
            if (err.response?.status !== 404) {
                error.value = 'Failed to fetch schedule'
                throw err
            }
            return null
        } finally {
            loading.value = false
        }
    }

    // Create or update schedule
    async function saveSchedule(workflowId: string, data: Partial<HealthCheckSchedule>) {
        loading.value = true
        error.value = null
        try {
            const response = await workflowsApi.createSchedule(workflowId, data)
            schedules.value.set(workflowId, response.data)
            return response.data
        } catch (err) {
            error.value = 'Failed to save schedule'
            throw err
        } finally {
            loading.value = false
        }
    }

    // Delete schedule
    async function deleteSchedule(workflowId: string) {
        loading.value = true
        error.value = null
        try {
            await workflowsApi.deleteSchedule(workflowId)
            schedules.value.delete(workflowId)
        } catch (err) {
            error.value = 'Failed to delete schedule'
            throw err
        } finally {
            loading.value = false
        }
    }

    // Test notification
    async function testNotification(workflowId: string, config: NotificationConfig) {
        loading.value = true
        error.value = null
        try {
            await workflowsApi.testNotification(workflowId, config)
        } catch (err) {
            error.value = 'Failed to send test notification'
            throw err
        } finally {
            loading.value = false
        }
    }

    return {
        schedules,
        loading,
        error,
        getSchedule,
        fetchSchedule,
        saveSchedule,
        deleteSchedule,
        testNotification
    }
})
